### PR TITLE
[Auditor] Add pass to make all binary files executable

### PR DESCRIPTION
**Pull request by giordano**
Related to https://github.com/JuliaLang/julia/issues/38993.  I verified locally
it works as I expect and will add tests later, after getting some feedback.

### CODE REVIEW

File: src/Auditor.jl [hunk 1]
```diff
@@ -85,6 +85,13 @@ function audit(prefix::Prefix, src_name::AbstractString = "";
                     all_ok &= check_isa(oh, platform, prefix; verbose=verbose, silent=silent)
                     # Check that the OS ABI is set correctly (often indicates the wrong linker was used)
                     all_ok &= check_os_abi(oh, platform, verbose = verbose)
+                    # Make sure all binary files are executables, if libraries aren't
+                    # executables Julia may not be able to dlopen them:
+                    # https://github.com/JuliaLang/julia/issues/38993.  In principle this
+                    # should be done when autofix=true, but we have to run this fix on MKL
+                    # for Windows, for which however we have to set autofix=false:
```
Comment (staticfloat) [inline src/Auditor.jl:92]: Another reason why we should not hide such complex functionality behind a simple `autofix` flag, and we should support enabling/disabling of each audit pass separately. :)

File: src/auditor/filesystems.jl [hunk 2]
```diff
@@ -43,3 +43,22 @@ function check_absolute_paths(prefix::Prefix, all_files::Vector; silent::Bool =
 
     return true
 end
+
+function ensure_executability(oh::ObjectHandle; verbose::Bool=false, silent::Bool=false)
+    if filemode(path(oh)) & 0o755 != 0o755
```
Comment (staticfloat) [inline src/auditor/filesystems.jl:48]: Instead of checking for exactly `0o755`, I think we might instead want to check `filemode(path(oh)) & 0o111 == 0`, and then we probably want to set `+x` for any groups that have `+r`; so we want to set `new_mode = old_mode | (old_mode & 0o444 >> 2)`.

File: src/Auditor.jl (hunk 1 above)
Comment (giordano) [reply src/Auditor.jl:92]: Agreed!

File: src/auditor/filesystems.jl (hunk 2 above)
Comment (giordano) [reply src/auditor/filesystems.jl:48]: I knew I was doing it wrong, I was too tired yesterday night :grimacing: Thanks!

Comment (giordano) [reply src/auditor/filesystems.jl]: I'm not sure I understand why `old_mode | (old_mode & 0o444 >> 2)`:
```julia
julia> old_mode = 0o644
0x01a4

julia> string(old_mode | (old_mode & 0o444 >> 2) ; base=8)
"644"
```
doesn't look right.  I think it's just
```julia
julia> string(old_mode | 0o111; base=8)
"755"
```
no?

Comment (staticfloat) [reply src/auditor/filesystems.jl]: you need more parens:

```
julia> old_mode = 0o640
0x01a0

julia> string(old_mode | ((old_mode & 0o444) >> 2); base=8)
"750"
```

Comment (staticfloat) [reply src/auditor/filesystems.jl]: My bad.  :P

Comment (giordano) [reply src/auditor/filesystems.jl]: `750` still doesn't have exec bit for others... or am I missing something?  Do you want to require readability in order to add the exec permission?  `chmod +x` just adds `x` everywhere

Comment (giordano) [reply src/auditor/filesystems.jl]: As weird as it is, it doesn't look like a file need to be readable, in order to execute it:
```
% cat foo.c 
#include <stdio.h>

int main()
{
    printf("Hello World!\n");
    return 0;
}
% gcc -o foo foo.c
% chmod 100 foo   
% ./foo 
Hello World!
```

Comment (staticfloat) [reply src/auditor/filesystems.jl]: Yeah, that is weird.

That being said, what I want is for us to catch obvious mistakes (e.g. `mode & 0o111 == 0`) and to fix them without accidentally giving permissions to users that it wasn't intended for.  For instance, if we have SETUID binaries in an `/sbin` folder that accidentally didn't have permissions applied (very niche scenario, I know) we wouldn't want to turn `0o600` into `0o711`; we'd want to turn it into `0o700`.

Comment (giordano) [reply src/auditor/filesystems.jl]: All right, I applied the change suggested!

File: test/auditing.jl [hunk 3]
```diff
@@ -517,6 +517,49 @@ end
...
+        tarball_path, tarball_hash = build_output_meta[platform][1:2]
+        @test isfile(tarball_path)
+
+        # Unpack it somewhere else
+        @test verify(tarball_path, tarball_hash)
+        testdir = joinpath(build_path, "testdir")
+        mkdir(testdir)
+        unpack(tarball_path, testdir)
+        libfoo_path = joinpath(testdir, build_output_meta[platform][4][product]["path"])
+        @test_broken filemode(libfoo_path) & 0o777 == 0o750
```
Comment (giordano) [inline test/auditing.jl:559]: I believe we have an issue when packing the tarball: the file has permissions `0x750` as expected right after `ensure_executability`, but in the tarball it becomes `0o755`.  However this is orthogonal to this PR

Comment (giordano) [reply test/auditing.jl:559]: It seems to be a deliberate decision in `Tar.jl`:

> files that are executable by the owner are archived with mode `0o755`;

I can reproduce this with
```julia
julia> using Tar

julia> mktempdir() do dir
           # Create the root of the tarball
           root = joinpath(dir, "root")
           mkpath(root)
           # Create the file and give permissions 750
           path = joinpath(root, "testfile")
           touch(path)
           chmod(path, 0o750)
           # Package the root
           tarball = Tar.create(root, joinpath(dir, "test.tar"))
           files = Tar.list(tarball)
           # Get the permissions of the file
           string(filter(f -> f.type === :file, files)[1].mode; base=8)
       end
"755"
```

Comment (staticfloat) [reply test/auditing.jl:562]: Yeah; I think it’s good for BB to be correct, and Tar is fine to do it’s thing as well.
